core/src/main/java/com/oath/oak/ThreadIndexCalculator.java

```diff
@@ -0,0 +1,35 @@
+
+package com.oath.oak;
+
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
+import java.util.concurrent.atomic.AtomicInteger;
+
+public class ThreadIndexCalculator {
+
+    private final AtomicInteger nextFreeIndex;
+    private final ConcurrentMap<Long, Integer> map;
+    static final int MAX_THREADS = 32;
+
+    private ThreadIndexCalculator() {
+        this.nextFreeIndex = new AtomicInteger(0);
+        map = new ConcurrentHashMap<>();
+    }
+
+
+    public int getIndex() {
+        long tid = Thread.currentThread().getId();
```

Comment:
I don't understand why is this solution correct or necessary.
Are we only trying to change the order of the indices?
I thought we had to maintain a set of "available" indices, which means a adding and removing items from the set.
Right now it seems that once all thread ask for their index it is permanent.
So why not using a simple modulo operation?